src/main/java/org/openstreetmap/atlas/utilities/collections/ShardBucketCollection.java

```diff
@@ -0,0 +1,454 @@
+        this.collectionIndex = new RTree<>();
+        final Counter counter = new Counter();
+        shards.forEach(shardBucket ->
+        {
+            final ShardToCollectionIndex shardToCollectionIndex = new ShardToCollectionIndex(
+                    (int) counter.getValueAndIncrement(), shardBucket);
+            this.collectionIndex.add(shardToCollectionIndex.bounds(), shardToCollectionIndex);
+        });
+        this.collectionBuckets = (CollectionType[]) Array.newInstance(
+                this.initializeBucketCollection().getClass(), (int) counter.getValue());
+    }
+
+    @Override
+    public final boolean add(final LocatedType item)
+    {
+        if (Objects.nonNull(item) && item.bounds().overlaps(this.maximumBounds))
+        {
+            final List<ShardToCollectionIndex> indexes = this.collectionIndex.get(item.bounds());
+            if (indexes.size() == 1)
+            {
+                return this.addFunction(item, this.getOrCreateBucketCollectionAt(indexes.get(0)),
+                        indexes.get(0).getShard());
+            }
+            else if (this.allowMultipleBucketInsertion())
+            {
+                final long addedAmount = indexes.stream()
+                        .filter(index -> this.addFunction(item,
+                                this.getOrCreateBucketCollectionAt(index), index.getShard()))
+                        .count();
+                return addedAmount > 0;
+            }
+            else
+            {
+                final Shard toInsertAt = this.resolveShard(item, indexes.stream()
+                        .map(ShardToCollectionIndex::getShard).collect(Collectors.toList()));
+                final Optional<ShardToCollectionIndex> toAddTo = indexes.stream()
+                        .filter(index -> toInsertAt.equals(index.getShard())).findFirst();
+                return toAddTo
+                        .map(index -> this.addFunction(item,
+                                this.getOrCreateBucketCollectionAt(index), index.getShard()))
+                        .orElse(false);
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean addAll(final Collection<? extends LocatedType> collection)
+    {
+        if (Objects.nonNull(collection))
+        {
+            final long addCount = collection.stream().filter(this::add).count();
+            return addCount > 0;
+        }
+        return false;
+    }
+
+    @Override
+    public void clear()
+    {
+        synchronized (this.collectionBuckets)
+        {
+            for (int i = 0; i < this.collectionBuckets.length; i++)
+            {
+                this.collectionBuckets[i] = null;
+            }
+            this.initializedShards.clear();
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean contains(final Object object)
+    {
+        final Optional<LocatedType> typedItem = this.castToLocatedType(object);
+        if (typedItem.isPresent())
```

Comment:
I like this pattern a lot actually. I will be stealing it ;)

Reply:
credit due to @matthieun